src/common/utils.lua

```diff
@@ -241,47 +233,78 @@ function utils.camelCase(str)
 	return str
 end
 
-utils.notices = {}
-
-function utils.notices.credit(msg)
-	local notice = {
-		Target = msg.From,
-		Action = "Credit-Notice",
-		Recipient = msg.Recipient,
-		Quantity = tostring(1),
-	}
-	for tagName, tagValue in pairs(msg) do
-		-- Tags beginning with "X-" are forwarded
-		if string.sub(tagName, 1, 2) == "X-" then
-			notice[tagName] = tagValue
+function utils.indexOf(t, value)
+	for i, v in ipairs(t) do
+		if v == value then
+			return i
+		end
+	end
+	return -1
+end
+
+function utils.updateUserAssociations(antId, state, user)
+	-- remove previous associations
+	if USERS[user] and user ~= state.Owner and not utils.includes(user, state.Controllers) then
+		local antIndex = utils.indexOf(USERS[user], antId)
+		if antIndex then
+			table.remove(USERS[user], antIndex)
 		end
 	end
 
-	return notice
+	-- add new associations
+	if user == state.Owner or utils.includes(user, state.Controllers) then
+		if not USERS[user] then
+			USERS[user] = {}
+		end
+		if not utils.includes(antId, USERS[user]) then
+			table.insert(USERS[user], antId)
+		end
+	end
 end
 
-function utils.notices.debit(msg)
-	local notice = {
-		Target = msg.From,
-		Action = "Debit-Notice",
-		Recipient = msg.Recipient,
-		Quantity = tostring(1),
-	}
-	-- Add forwarded tags to the credit and debit notice messages
-	for tagName, tagValue in pairs(msg) do
-		-- Tags beginning with "X-" are forwarded
-		if string.sub(tagName, 1, 2) == "X-" then
-			notice[tagName] = tagValue
+function utils.updateAssociations(antId, state)
+	-- Remove previous associations for old owner and controllers
+	local previousOwner = ANTS[antId].Owner
+	local previousControllers = ANTS[antId].Controllers
+
+	if previousOwner and previousOwner ~= state.Owner and type(ANTS[antId]) == "table" then
```

Comment:
If you use the object structures I advocated for, this just becomes:
```
-- If the previous owner is no longer associated with this ANT, remove their association to it
if previousOwner and previousOwner ~= state.Owner and state.Controllers[previousOwner] == nil then
  ADDRESSES[previousOwner][antId] = nil
end
```
This might require mapping to new ant state into the format that we're now storing it first and passing into this function that way.

You can apply this type of logic and coding style everywhere else in this PR.

Reply:
addresses[previousOwner][antId] = true and utils.isAffiliated(previousOwner, newAnt) or nil
for prevController, _ in pairs(ANTS[antId].Controllers) do
    addresses[prevController][antId] = true and utils.isAffiliated(prevController, newAnt) or nil
end


for controller, _ in pairs(newAnt.Controllers) do
    addresses[controller][antId] = newAnt
end
addresses[newAnt.Owner][antId] = newAnt